segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -120,295 +139,572 @@
+            throw new CompletionException(new StorageMetadataException("Exception during execution of external commit step", e));
+        }
+        return CompletableFuture.completedFuture(null);
+    }
 
-            // Execute external commit step.
-            try {
-                if (null != txn.getExternalCommitStep()) {
-                    txn.getExternalCommitStep().call();
-                }
-            } catch (Exception e) {
-                log.error("Exception during execution of external commit step", e);
-                throw new StorageMetadataException("Exception during execution of external commit step", e);
+    private void validateCommit(MetadataTransaction txn, Map<String, TransactionData> txnData, ArrayList<String> modifiedKeys, ArrayList<TransactionData> modifiedValues) {
+        for (val entry : txnData.entrySet()) {
+            val key = entry.getKey();
+            val transactionData = entry.getValue();
+            Preconditions.checkState(null != transactionData.getKey());
+
+            // See if this entry was modified in this transaction.
+            if (transactionData.getVersion() == txn.getVersion()) {
+                modifiedKeys.add(key);
+                transactionData.setPersisted(false);
+                modifiedValues.add(transactionData);
             }
+            // make sure none of the keys used in this transaction have changed.
+            val dataFromBuffer = bufferedTxnData.get(key);
+            if (null != dataFromBuffer) {
+                if (!dataFromBuffer.isPinned()) {
+                    Preconditions.checkState(null != dataFromBuffer.getDbObject());
+                }
+                if (dataFromBuffer.getVersion() > transactionData.getVersion()) {
+                    throw new CompletionException(new StorageMetadataVersionMismatchException(
+                            String.format("Transaction uses stale data. Key version changed key:%s buffer:%s transaction:%s",
+                                    key, dataFromBuffer.getVersion(), txnData.get(key).getVersion())));
+                }
+
+                // Pin it if it is already pinned.
+                transactionData.setPinned(transactionData.isPinned() || dataFromBuffer.isPinned());
 
-            // If we reach here then it means transaction is safe to commit.
-            // Step 4: Insert
-            long committedVersion = version.incrementAndGet();
-            HashMap<String, TransactionData> toAdd = new HashMap<String, TransactionData>();
-            for (String key : modifiedKeys) {
-                TransactionData data = txnData.get(key);
-                data.setVersion(committedVersion);
-                toAdd.put(key, data);
+                // Set the database object.
+                transactionData.setDbObject(dataFromBuffer.getDbObject());
+            } else {
+                Preconditions.checkState(entry.getValue().isPinned(), "Why are we here??");
```

Comment:
Yes. Why, oh why are we here? 

(How is this useful in a production log?)

Reply:
:) fixed